### ISSUE TITLE
addjunction ignores existence of tanks/reservoirs

### ISSUE THREAD

**Issue by samhatchett** [OpenWaterAnalytics/EPANET#542]
Did we know about this already? Issue related to the network modification api. If we create a project, and add a junction, then a tank... all is well. But then adding another junction overwrites the tank and leaves an incomplete `Snode` struct hanging off the end of the Node list. Closing the project handle results in a double-free crash.

https://github.com/OpenWaterAnalytics/EPANET/blob/5f7fd55a36e6924e95a6ec9e2bf327decb3a8935/src/epanet.c#L1746-L1751

This logic assumes that no tanks/reservoirs are present. The solution would seem to be another function which goes to the end of the Node list and copies all the node data +1 index for all indexes >= `Njuncs`, thus leaving a clean destination for this new node.

The same type of issue probably also exists in `addlink`

**Comment by samhatchett** [comment]
I suppose either an additional function, or since there is already considerable "index-shifting" going on within this function, we could add it here. Also "note to self": the `Snode.Type` field is not currently set in `addnode`

**Comment by LRossman** [comment]
Wouldn't the simplest fix be to move the current code segment that begins at `// shift rest of Node array `to before the` net->Njuncs++` line as shown below (with a small change to move only the Tank/Reservoir nodes):
```
    if (nodeType == EN_JUNCTION)
    {
        // shift indices of non-Junction nodes at end of Node array
        for (i = net->Nnodes; i > net->Njuncs; i--)
        {
            hashtable_update(net->NodeHashTable, net->Node[i].ID, i + 1);
            net->Node[i + 1] = net->Node[i];
        }
        net->Njuncs++;
        nIdx = net->Njuncs;
        node = &net->Node[nIdx];
        node->D = NULL;
        adddemand(node, 0.0, 0, NULL);

        // shift indices of Tank array
```

**Comment by LRossman** [comment]
I've provided a fix for this in PR #543. It also assigns a value for the `Snode.Type` field and uses a modified version of the `test_node.cpp` unit test to check that the fix works. After examining the `EN_addlink()` function I can't see that a similar problem exists there since new links are always appended to the end of the `Network.Link` array no matter what type they are.

**Comment by samhatchett** [comment]
thanks @LRossman - I like the solution. So that seems a bit surprising about the Link situation - but looking at the code it makes sense.
